contrib/win32/win32compat/fileio.c

```diff
@@ -1106,39 +1044,159 @@ fileio_readlink(const char *path, char *buf, size_t bufsiz)
 		goto cleanup;
 	}
 
-	/* read the link data from the passed symlink */
-	resolved_link = fileio_readlink_internal(wpath);
-	if (resolved_link == NULL) {
-		/* errorno will have been set by called function */
+	/* obtain a handle to send to deviceioctl  */
+	handle = CreateFileW(wpath, 0, 0, NULL, OPEN_EXISTING, 
+		FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OPEN_REPARSE_POINT, 0);
+	if (handle == INVALID_HANDLE_VALUE) {
+		errno = errno_from_Win32LastError();
+		goto cleanup;
+	}
+
+	/* send a request to the file system to get the real path */
+	reparse_buffer = (PREPARSE_DATA_BUFFER_SYMLINK) malloc(MAXIMUM_REPARSE_DATA_BUFFER_SIZE);
+	DWORD dwBytesReturned = 0;
+	if (DeviceIoControl(handle, FSCTL_GET_REPARSE_POINT, NULL, 0,
+		(LPVOID) reparse_buffer, MAXIMUM_REPARSE_DATA_BUFFER_SIZE, &dwBytesReturned, 0) == 0) {
+		errno = errno_from_Win32LastError();
 		goto cleanup;
 	}
 
-	if ((output = utf16_to_utf8(resolved_link)) == NULL) {
-		errno = ENOMEM;
+	/* ensure file is actually symlink */
+	if (reparse_buffer->ReparseTag != IO_REPARSE_TAG_SYMLINK) {
+		errno = EINVAL;
+		goto cleanup;
+	}
+
+	/* the symlink structure has a 'Print Name' value that is displayed to the
+	 * user which is different from the actual value it uses for redirection
+	 * called the 'Substitute Name'; since the Substitute Name has an odd format
+	 * that begins with \??\ and it appears that CreateSymbolicLink() always
+	 * formats the PrintName value consistently we will just use that
+	 */
+	int symlink_nonnull_size = reparse_buffer->PrintNameLength;
+	wchar_t * symlink_nonnull = &reparse_buffer->PathBuffer[reparse_buffer->PrintNameOffset / sizeof(WCHAR)];
+
+	/* allocate area to hold a null terminated version of the string */
+	if ((linkpath = malloc(symlink_nonnull_size + sizeof(wchar_t))) == NULL) {
```

Comment:
I think it should be like
malloc(symlink_nonnull_size + 1 + sizeof(wchar_t)); // add 1 for NULL character.

Reply:
This is a unicode (wide char) string so that's actually what the sizeof(wchar_t) is used for.  The sizeof(wchar_t) is actually equal to 2.  The two additional bytes is for the 2-byte null character.